### ISSUE TITLE
Changing prefetch and preload options (resourceHints)

### ISSUE THREAD

**Issue by tmorehouse** [nuxt/nuxt.js#1838]
Currently under `the` `render` section of `nuxt.config.js` there is an option `resourceHints` (boolean), which enabled (default) or disabled (by setting to `false`)  which, based on the documentation:

> Adds prefetch and preload links for faster initial page load time.

I have a rather large list of routes, and some routes that are only used by users that have certain roles/privileges.

It would be nice if:

- `prefetch` could be disabled, while still having `preload` (for things like the common chunks) enabled.
- Set a flag in the pages (similar to how `scrollToTop` is flagged), to enable/disble `preload` or, `prefetch`. for that particular page component
- Specify chunknames (or grouping) for a set of routes (i.e. via the page component definition) to group several small pages into the same webpackChunkName file

99% of the my users cannot access `administrative` pages, so it is kind of a waste of b/w to pre-fetch those pages which will never be loaded by those users, without having to disable site-wide pre-fetching (which affects 100% of the users)

An other option would be able to group pre-fetches by a user speified parameter... say if someone accesses /admin (and isn't re-directed/blocked by middleware), then all the child routes of `/admin would be listed/added in the head as prefetch links.

<div align="right"><sub><em>This feature request is available on <a href="https://nuxtjs.cmty.io">Nuxt.js</a> community (<a href="https://nuxtjs.cmty.io/nuxt/nuxt.js/issues/c1652">#c1652</a>)</em></sub></div>

**Comment by Atinux** [comment]
Hi @tmorehouse 

This is actually a nice idea but not so easy to implement, I am talking with Vue.js core-team to see how they could implement something like this into the core of `vue-server-renderer` since it's this module that handle the `resourceHints`.

**Comment by tmorehouse** [comment]
I see that vue-server-renderer has teh following options:
- shouldPreload https://ssr.vuejs.org/en/api.html#shouldpreload
- shouldPrefetch https://ssr.vuejs.org/en/api.html#shouldprefetch

And both hooks are passed a filename and type.  Is there a way to configure these?

I could use these to test the filename (which is based on the page path) to signal a true/false to have certain paths not included in the prefetch list?

**Comment by Atinux** [comment]
@tmorehouse 

Actually you can customize it into your `nuxt.config.js`, see my answer on this issue: https://github.com/nuxt/nuxt.js/issues/1508#issuecomment-325655533

**Comment by Atinux** [comment]
Hi @tmorehouse 

It's really weird, I let this issue open, we will see with next release since I'm changing a few things right now.

**Comment by tmorehouse** [comment]
Hi @Atinux 

Just wondering if you have discovered anything about this?

**Comment by Atinux** [comment]
Hey @tmorehouse 

I am focusing on releasing 1.0 right now, I added the 2.0 label here (but it could be implemented right after 1.0 with 1.1 for example).

I like your idea and might need more digging, that's why I keep it open :)

**Comment by buhrmi** [comment]
Hello, guys. I have the same problem. I want to include some pages in the resourceHint list but others I don't want to include. I think it would be a good solution to have the setting in the page component directly, eg:

```
<script>
export default {
  data: {..},
  preload: false
}
</script>
```

Let me know what you think.

**Comment by tmorehouse** [comment]
It would also be nice to be able to see this based on some runtime condition as well, such as user roles, etc.

**Comment by buhrmi** [comment]
@tmorehouse i think the browser resource hints are generated at compile time, so that would be impossible.

**Comment by tmorehouse** [comment]
Compile time, but stored in the manifest.

**Comment by randyhoulahan** [comment]
I missed https://nuxtjs.org/api/configuration-render#resourcehints can disable for large applications.

**Comment by manniL** [comment]
Update on that issue. I've tried using `shouldPrefetch` and it worked quite well as described above:

```
  render: {
    bundleRenderer: {
      shouldPrefetch: (file, type) => {
         if (type === 'script') {
            if (/yourPageNameHere/.test(file)) {
               return false
            }
         }
         return true
      }
    }
  }
```

**Comment by aadii104** [comment]
hey @manniL,
could you please tell about how and which file I should add here in this code,
```js
 if (/yourPageNameHere/.test(file)) {
               return false
            }
```
and do u have any repo for this working?
thanks.

**Comment by manniL** [comment]
Hey @aadii104 :relaxed: 

You need to include the **whole** code snippet from my comment above in your `nuxt.config.js`.

I'm using custom prefetch/preload for my own website. You can find the related code [here](https://github.com/manniL/lichter.io/blob/473bcc19203dda9c34933e804d3cb661edbe1678/nuxt.config.js#L116).

Be aware that this solution does not work with `nuxt-edge` by default.

**Comment by sschadwick** [comment]
@manniL Why does it not work with `nuxt-edge`? Trying to solve this same issue on `nuxt-edge@latest`

**Comment by manniL** [comment]
@sschadwick The functions behavior actually hasn't changed in `nuxt-edge`. Be aware that `shouldPrefetch` defaults to `() => false` while `shouldPreload` defaults to `() => true`.

But as files aren't named by default, you'll have to enable it (in case you want to distinguish by name and not by type). You can reenable names with:

nuxt.config.js
```js
export default {
  build: {
    optimization: {
      splitChunks: {
        name: true
      }
    }
  }
}
```

**Comment by manniL** [comment]
@sschadwick I use the same approach for my blog. Here is a part of my nuxt.config.js 

```js
export default {
  /*
  * Render (preload & prefetch)
  */
  render: {
    bundleRenderer: {
      shouldPrefetch: (file, type) => ['script', 'style', 'font'].includes(type) && !file.includes('admin')
    }
  },

  /*
   * Build configuration
   */
  build: {
    optimization: {
      splitChunks: {
        name: true
      }
    }
```

**Comment by sschadwick** [comment]
@manniL I took a brief look at your blog code, and saw it was working as intended. However, when I update `nuxt-edge` to latest, the naming structure of the bundles changes possibly to the point where `!file.includes('admin')` no longer matches the protected routes. I'm curious if updating your deps would break the functionality.

When I use `import('./component.vue')` to lazy load components, they are each distributed as separate files. The only way that I am able to bundle all of the protected routes together is to use a different `splitChunks` object and not lazy import. However, this ceases to respect the `shouldPrefetch`  function. I can confirm this by setting `shouldPrefetch` to always return false, yet the `admin` bundle will still be served on initial page load. Ideally it would not be served until an admin user clicks an admin route. Any ideas?

```javascript
module.exports = {
  build: {
    optimization: {
      splitChunks: {
        cacheGroups: {
          admin: {
            chunks:   'all',
            priority: 1,
            test:     <REGEX TO MATCH .VUE FILES IN PROTECTED DIR>
          }
        }
      }
    }
  },
}
```

**Comment by sschadwick** [comment]
I was finally able to configure my build exactly how I wanted. Here's some lessons I learned:
* Lazy load with `() => import('./Component')` instead of `import('./Component')`.
* There were also a couple of babel issues, so I had to ultimately use `() => import('./Component').then(m => m.default || m));`
* There were two types of JS files in my protected routes: a text abstraction and a routes abstraction, so I had to update my webpack regex to include everything but the routes abstraction.
* I was unable to get `shouldPreload` or `shouldPrefetch` to work as expected, so I ultimately set `resourceHints: false`.

To protect the specified-routes bundle:
* Enable build filenames, as demonstrated above.
* Check the url for the name of the protected bundle before Nuxt is passed to Express. Deny bundle if unauthenticated/unauthorized.
* The internal Nuxt serverMiddleware does not receive requests for any files, only in-app paths.
